File: implementations/java/server/src/main/java/com/theshuai/specusserver/management/service/HttpMediaManifestSupport.java
package com.theshuai.specusserver.management.service;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class HttpMediaManifestSupport {
    static final String HLS_MANIFEST = "HLS_MANIFEST";
    static final String DASH_MANIFEST = "DASH_MANIFEST";
    static final String PROGRESSIVE = "PROGRESSIVE";
    static final String MEDIA_SEGMENT = "MEDIA_SEGMENT";

    private static final Pattern HLS_URI_ATTRIBUTE = Pattern.compile("URI=(\"([^\"]+)\"|'([^']+)')");
    private static final Pattern DASH_URI_ATTRIBUTE = Pattern.compile(
            "(?i)(media|initialization|sourceURL|href)\\s*=\\s*(\"([^\"]+)\"|'([^']+)')");
    private static final Pattern DASH_BASE_URL = Pattern.compile(
            "(?is)<BaseURL(\\s[^>]*)?>([^<]+)</BaseURL>");
    private static final Pattern CONTENT_RANGE = Pattern.compile(
            "(?i)^bytes\\s+(\\d+)-(\\d+)/(\\d+|\\*)$");
    private static final Pattern NUMBER_IN_NAME = Pattern.compile("(\\d+)(?=\\.[^.]+$)");

    private HttpMediaManifestSupport() {
    }

    static String classify(String sourceUrl, String contentType, int statusCode, String contentRange) {
        String type = mediaType(contentType);
        String path = pathOnly(sourceUrl).toLowerCase(Locale.ROOT);
        if (type.equals("application/vnd.apple.mpegurl")
                || type.equals("application/x-mpegurl")
                || type.equals("audio/mpegurl")
                || path.endsWith(".m3u8")) {
            return HLS_MANIFEST;
        }
        if (type.equals("application/dash+xml") || path.endsWith(".mpd")) {
            return DASH_MANIFEST;
        }
        if (isSegmentPath(path) || type.equals("video/mp2t")
                || type.equals("application/octet-stream")) {
            return MEDIA_SEGMENT;
        }
        if (type.startsWith("video/") || type.startsWith("audio/")
                || isProgressivePath(path)
                || (statusCode == 206 && contentRange != null)) {
            return PROGRESSIVE;
        }
        return null;
    }

    static ContentRange parseContentRange(String value) {
        if (value == null) {
            return null;
        }
        Matcher matcher = CONTENT_RANGE.matcher(value.trim());
        if (!matcher.matches()) {
            return null;
        }
        long start = Long.parseLong(matcher.group(1));
        long end = Long.parseLong(matcher.group(2));
        Long total = "*".equals(matcher.group(3)) ? null : Long.parseLong(matcher.group(3));
        if (end < start || (total != null && end >= total)) {
            return null;
        }
        return new ContentRange(start, end, total);
    }

    static Long inferSequence(String sourceUrl) {
        Matcher matcher = NUMBER_IN_NAME.matcher(pathOnly(sourceUrl));
        if (!matcher.find()) {
            return null;
        }
        try {
            return Long.parseLong(matcher.group(1));
        } catch (NumberFormatException ignored) {
            return null;
        }
    }

    static boolean isInitializationSegment(String sourceUrl) {
        String path = pathOnly(sourceUrl).toLowerCase(Locale.ROOT);
        String file = path.substring(path.lastIndexOf('/') + 1);
        return file.startsWith("init.") || file.startsWith("init-") || file.contains("initialization");
    }

    static ParsedManifest parse(String kind, String sourceUrl, String text) {
        if (HLS_MANIFEST.equals(kind)) {
            return parseHls(sourceUrl, text);
        }
        if (DASH_MANIFEST.equals(kind)) {
            return parseDash(sourceUrl, text);
        }
        return new ParsedManifest(false, List.of());
    }

    static String rewrite(String kind, String sourceUrl, String text, String assetBasePath) {
        if (HLS_MANIFEST.equals(kind)) {
            return rewriteHls(sourceUrl, text, assetBasePath);
        }
        if (DASH_MANIFEST.equals(kind)) {
            return rewriteDash(sourceUrl, text, assetBasePath);
        }
        return text;
    }

    static String resolveSourceUrl(String baseSourceUrl, String reference) {
        if (reference == null || reference.isBlank() || reference.startsWith("data:")) {
            return reference;
        }
        try {
            URI base = URI.create("https://capture.invalid" + normalizeSourceUrl(baseSourceUrl));
            URI resolved = base.resolve(reference.trim());
            String path = resolved.getRawPath();
            if (path == null || path.isBlank()) {
                path = "/";
            }
            return normalizeSourceUrl(path + (resolved.getRawQuery() == null ? "" : "?" + resolved.getRawQuery()));
        } catch (IllegalArgumentException ignored) {
            return normalizeSourceUrl(reference);
        }
    }

    static String normalizeSourceUrl(String value) {
        if (value == null || value.isBlank()) {
            return "/";
        }
        String normalized = value.trim();
        try {
            URI uri = URI.create(normalized);
            if (uri.isAbsolute()) {
                normalized = (uri.getRawPath() == null || uri.getRawPath().isBlank() ? "/" : uri.getRawPath())
                        + (uri.getRawQuery() == null ? "" : "?" + uri.getRawQuery());
            }
        } catch (IllegalArgumentException ignored) {
            // Keep the original relative value and normalize it below.
        }
        if (!normalized.startsWith("/")) {
            normalized = "/" + normalized;
        }
        return normalized.length() <= 3072 ? normalized : normalized.substring(0, 3072);
    }

    private static ParsedManifest parseHls(String sourceUrl, String text) {
        List<ManifestReference> references = new ArrayList<>();
        long mediaSequence = 0;
        boolean endList = false;
        boolean hasMediaSegment = false;
        String[] lines = text.split("\\R", -1);
        for (String line : lines) {
            String trimmed = line.trim();
            if (trimmed.startsWith("#EXT-X-MEDIA-SEQUENCE:")) {
                try {
                    mediaSequence = Long.parseLong(trimmed.substring(trimmed.indexOf(':') + 1).trim());
                } catch (NumberFormatException ignored) {
                    mediaSequence = 0;
                }
            } else if (trimmed.equals("#EXT-X-ENDLIST")) {
                endList = true;
            }

            Matcher attribute = HLS_URI_ATTRIBUTE.matcher(line);
            while (attribute.find()) {
                String uri = firstText(attribute.group(2), attribute.group(3));
                String relation = trimmed.startsWith("#EXT-X-MAP") ? "INITIALIZATION"
                        : trimmed.startsWith("#EXT-X-KEY") ? "KEY" : "ASSET";
                references.add(new ManifestReference(relation, null, uri, resolveSourceUrl(sourceUrl, uri)));
            }
            if (!trimmed.isEmpty() && !trimmed.startsWith("#")) {
                String relation = pathOnly(trimmed).toLowerCase(Locale.ROOT).endsWith(".m3u8")
                        ? "PLAYLIST" : "SEGMENT";
                hasMediaSegment = hasMediaSegment || "SEGMENT".equals(relation);
                Long sequence = "SEGMENT".equals(relation) ? mediaSequence++ : null;
                references.add(new ManifestReference(
                        relation, sequence, trimmed, resolveSourceUrl(sourceUrl, trimmed)));
            }
        }
        return new ParsedManifest(hasMediaSegment && !endList, references);
    }

    private static ParsedManifest parseDash(String sourceUrl, String text) {
        List<ManifestReference> references = new ArrayList<>();
        Matcher attributes = DASH_URI_ATTRIBUTE.matcher(text);
        long sequence = 0;
        while (attributes.find()) {
            String name = attributes.group(1).toLowerCase(Locale.ROOT);
            String uri = firstText(attributes.group(3), attributes.group(4));
            String relation = name.equals("initialization") || name.equals("sourceurl")
                    ? "INITIALIZATION" : "SEGMENT";
            references.add(new ManifestReference(
                    relation,
                    "SEGMENT".equals(relation) ? sequence++ : null,
                    uri,
                    resolveSourceUrl(sourceUrl, uri)));
        }
        Matcher baseUrls = DASH_BASE_URL.matcher(text);
        while (baseUrls.find()) {
            String uri = baseUrls.group(2).trim();
            if (!uri.isBlank()) {
                references.add(new ManifestReference(
                        "BASE", null, uri, resolveSourceUrl(sourceUrl, uri)));
            }
        }
        boolean live = Pattern.compile("(?i)<MPD\\b[^>]*\\btype\\s*=\\s*[\"']dynamic[\"']")
                .matcher(text).find();
        return new ParsedManifest(live, references);
    }

    private static String rewriteHls(String sourceUrl, String text, String assetBasePath) {
        StringBuilder result = new StringBuilder(text.length() + 256);
        String[] lines = text.split("\\R", -1);
        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];
            Matcher matcher = HLS_URI_ATTRIBUTE.matcher(line);
            StringBuffer attributes = new StringBuffer();
            while (matcher.find()) {
                String original = firstText(matcher.group(2), matcher.group(3));
                String replacement = "URI=\"" + assetUrl(assetBasePath, resolveSourceUrl(sourceUrl, original)) + "\"";
                matcher.appendReplacement(attributes, Matcher.quoteReplacement(replacement));
            }
            matcher.appendTail(attributes);
            String rewritten = attributes.toString();
            String trimmed = rewritten.trim();
            if (!trimmed.isEmpty() && !trimmed.startsWith("#")) {
                rewritten = assetUrl(assetBasePath, resolveSourceUrl(sourceUrl, trimmed));
            }
            result.append(rewritten);
            if (i + 1 < lines.length) {
                result.append('\n');
            }
        }
        return result.toString();
    }

    private static String rewriteDash(String sourceUrl, String text, String assetBasePath) {
        Matcher attributes = DASH_URI_ATTRIBUTE.matcher(text);
        StringBuffer rewritten = new StringBuffer();
        while (attributes.find()) {
            String attributeName = attributes.group(1);
            String original = firstText(attributes.group(3), attributes.group(4));
            String value = xmlEscape(assetUrl(assetBasePath, resolveSourceUrl(sourceUrl, original)));
            attributes.appendReplacement(rewritten,
                    Matcher.quoteReplacement(attributeName + "=\"" + value + "\""));
        }
        attributes.appendTail(rewritten);

        Matcher baseUrls = DASH_BASE_URL.matcher(rewritten.toString());
        StringBuffer result = new StringBuffer();
        while (baseUrls.find()) {
            String prefix = baseUrls.group(1) == null ? "" : baseUrls.group(1);
            String value = xmlEscape(assetUrl(assetBasePath,
                    resolveSourceUrl(sourceUrl, baseUrls.group(2).trim())));
            baseUrls.appendReplacement(result,
                    Matcher.quoteReplacement("<BaseURL" + prefix + ">" + value + "</BaseURL>"));
        }
        baseUrls.appendTail(result);
        return result.toString();
    }

    private static String assetUrl(String assetBasePath, String resolvedSourceUrl) {
        String encoded = URLEncoder.encode(resolvedSourceUrl, StandardCharsets.UTF_8)
                .replace("+", "%20")
                .replace("%24", "$");
        return assetBasePath + "?url=" + encoded;
    }

    private static boolean isProgressivePath(String path) {
        return path.endsWith(".mp4") || path.endsWith(".webm") || path.endsWith(".mkv")
                || path.endsWith(".mov") || path.endsWith(".m4v") || path.endsWith(".mp3")
                || path.endsWith(".m4a") || path.endsWith(".ogg") || path.endsWith(".opus")
                || path.endsWith(".wav") || path.endsWith(".flac");
    }

    private static boolean isSegmentPath(String path) {
        return path.endsWith(".ts") || path.endsWith(".m4s") || path.endsWith(".cmfv")
                || path.endsWith(".cmfa") || path.endsWith(".aac") || path.endsWith(".vtt")
                || path.endsWith(".key");
    }

    private static boolean isLikelyMediaPath(String path) {
        return isProgressivePath(path) || isSegmentPath(path);
    }

    private static String pathOnly(String sourceUrl) {
        int query = sourceUrl == null ? -1 : sourceUrl.indexOf('?');
        return sourceUrl == null ? "" : query < 0 ? sourceUrl : sourceUrl.substring(0, query);
    }

    private static String mediaType(String contentType) {
        if (contentType == null) {
            return "";
        }
        return contentType.split(";", 2)[0].trim().toLowerCase(Locale.ROOT);
    }

    private static String firstText(String first, String second) {
        return first != null ? first : second;
    }

    private static String xmlEscape(String value) {
        return value.replace("&", "&amp;").replace("\"", "&quot;");
    }

    record ContentRange(long start, long end, Long total) {
    }

    record ManifestReference(
            String relationType,
            Long sequence,
            String originalUri,
            String resolvedSourceUrl
    ) {
    }

    record ParsedManifest(boolean live, List<ManifestReference> references) {
    }
}

File: implementations/csharp/server/src/Specus.Server/Management/HttpMediaManifestSupport.cs
using System.Net;
using System.Text.RegularExpressions;

namespace Specus.Server.Management;

internal static class HttpMediaManifestSupport
{
    public const string HlsManifest = "HLS_MANIFEST";
    public const string DashManifest = "DASH_MANIFEST";
    public const string Progressive = "PROGRESSIVE";
    public const string MediaSegment = "MEDIA_SEGMENT";

    private static readonly Regex HlsUriAttribute = new(
        "URI=(\\\"([^\\\"]+)\\\"|'([^']+)')", RegexOptions.Compiled);
    private static readonly Regex DashUriAttribute = new(
        "(media|initialization|sourceURL|href)\\s*=\\s*(\\\"([^\\\"]+)\\\"|'([^']+)')",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);
    private static readonly Regex DashBaseUrl = new(
        "<BaseURL(\\s[^>]*)?>([^<]+)</BaseURL>",
        RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Singleline);
    private static readonly Regex ContentRangePattern = new(
        "^bytes\\s+(\\d+)-(\\d+)/(\\d+|\\*)$",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);
    private static readonly Regex NumberInName = new(
        "(\\d+)(?=\\.[^.]+$)", RegexOptions.Compiled);
    private static readonly Regex DynamicMpd = new(
        "<MPD\\b[^>]*\\btype\\s*=\\s*[\"']dynamic[\"']",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    public static string? Classify(string sourceUrl, string? contentType, int statusCode,
        string? contentRange)
    {
        var type = MediaType(contentType);
        var path = PathOnly(sourceUrl).ToLowerInvariant();
        if (type is "application/vnd.apple.mpegurl" or "application/x-mpegurl" or "audio/mpegurl"
            || path.EndsWith(".m3u8", StringComparison.Ordinal))
        {
            return HlsManifest;
        }
        if (type == "application/dash+xml" || path.EndsWith(".mpd", StringComparison.Ordinal))
        {
            return DashManifest;
        }
        if (IsSegmentPath(path) || type is "video/mp2t" or "application/octet-stream")
        {
            return MediaSegment;
        }
        if (type.StartsWith("video/", StringComparison.Ordinal)
            || type.StartsWith("audio/", StringComparison.Ordinal)
            || IsProgressivePath(path) || statusCode == 206 && contentRange is not null)
        {
            return Progressive;
        }
        return null;
    }

    public static ContentRange? ParseContentRange(string? value)
    {
        if (value is null)
        {
            return null;
        }
        var match = ContentRangePattern.Match(value.Trim());
        if (!match.Success
            || !long.TryParse(match.Groups[1].Value, out var start)
            || !long.TryParse(match.Groups[2].Value, out var end))
        {
            return null;
        }
        long? total = null;
        if (match.Groups[3].Value != "*")
        {
            if (!long.TryParse(match.Groups[3].Value, out var parsedTotal) || parsedTotal < 0)
            {
                return null;
            }
            total = parsedTotal;
        }
        return end < start || total is not null && end >= total
            ? null
            : new ContentRange(start, end, total);
    }

    public static long? InferSequence(string sourceUrl)
    {
        var match = NumberInName.Match(PathOnly(sourceUrl));
        return match.Success && long.TryParse(match.Groups[1].Value, out var result) ? result : null;
    }

    public static bool IsInitializationSegment(string sourceUrl)
    {
        var path = PathOnly(sourceUrl).ToLowerInvariant();
        var file = path[(path.LastIndexOf('/') + 1)..];
        return file.StartsWith("init.", StringComparison.Ordinal)
               || file.StartsWith("init-", StringComparison.Ordinal)
               || file.Contains("initialization", StringComparison.Ordinal);
    }

    public static bool IsManifest(string? kind) => kind is HlsManifest or DashManifest;

    public static ParsedManifest Parse(string kind, string sourceUrl, string text) => kind switch
    {
        HlsManifest => ParseHls(sourceUrl, text),
        DashManifest => ParseDash(sourceUrl, text),
        _ => new ParsedManifest(false, []),
    };

    public static string Rewrite(string kind, string sourceUrl, string text, string assetBasePath) => kind switch
    {
        HlsManifest => RewriteHls(sourceUrl, text, assetBasePath),
        DashManifest => RewriteDash(sourceUrl, text, assetBasePath),
        _ => text,
    };

    public static string ResolveSourceUrl(string baseSourceUrl, string? reference)
    {
        if (string.IsNullOrWhiteSpace(reference) || reference.StartsWith("data:", StringComparison.Ordinal))
        {
            return reference ?? string.Empty;
        }
        try
        {
            var baseUri = new Uri("https://capture.invalid" + NormalizeSourceUrl(baseSourceUrl));
            var resolved = new Uri(baseUri, reference.Trim());
            return NormalizeSourceUrl(resolved.PathAndQuery);
        }
        catch (UriFormatException)
        {
            return NormalizeSourceUrl(reference);
        }
    }

    public static string NormalizeSourceUrl(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
        {
            return "/";
        }
        var normalized = value.Trim();
        if (Uri.TryCreate(normalized, UriKind.Absolute, out var absolute))
        {
            normalized = string.IsNullOrWhiteSpace(absolute.AbsolutePath)
                ? "/" + absolute.Query
                : absolute.PathAndQuery;
        }
        if (!normalized.StartsWith('/'))
        {
            normalized = "/" + normalized;
        }
        return normalized.Length <= 3072 ? normalized : normalized[..3072];
    }

    private static ParsedManifest ParseHls(string sourceUrl, string text)
    {
        var references = new List<ManifestReference>();
        long mediaSequence = 0;
        var endList = false;
        var hasMediaSegment = false;
        foreach (var line in Lines(text))
        {
            var trimmed = line.Trim();
            if (trimmed.StartsWith("#EXT-X-MEDIA-SEQUENCE:", StringComparison.Ordinal))
            {
                _ = long.TryParse(trimmed[(trimmed.IndexOf(':') + 1)..].Trim(), out mediaSequence);
            }
            else if (trimmed == "#EXT-X-ENDLIST")
            {
                endList = true;
            }

            foreach (Match match in HlsUriAttribute.Matches(line))
            {
                var uri = First(match.Groups[2].Value, match.Groups[3].Value);
                var relation = trimmed.StartsWith("#EXT-X-MAP", StringComparison.Ordinal)
                    ? "INITIALIZATION"
                    : trimmed.StartsWith("#EXT-X-KEY", StringComparison.Ordinal) ? "KEY" : "ASSET";
                references.Add(new ManifestReference(relation, null, uri,
                    ResolveSourceUrl(sourceUrl, uri)));
            }
            if (trimmed.Length > 0 && !trimmed.StartsWith('#'))
            {
                var relation = PathOnly(trimmed).EndsWith(".m3u8", StringComparison.OrdinalIgnoreCase)
                    ? "PLAYLIST" : "SEGMENT";
                hasMediaSegment |= relation == "SEGMENT";
                references.Add(new ManifestReference(relation,
                    relation == "SEGMENT" ? mediaSequence++ : null, trimmed,
                    ResolveSourceUrl(sourceUrl, trimmed)));
            }
        }
        return new ParsedManifest(hasMediaSegment && !endList, references);
    }

    private static ParsedManifest ParseDash(string sourceUrl, string text)
    {
        var references = new List<ManifestReference>();
        long sequence = 0;
        foreach (Match match in DashUriAttribute.Matches(text))
        {
            var name = match.Groups[1].Value.ToLowerInvariant();
            var uri = First(match.Groups[3].Value, match.Groups[4].Value);
            var relation = name is "initialization" or "sourceurl" ? "INITIALIZATION" : "SEGMENT";
            references.Add(new ManifestReference(relation, relation == "SEGMENT" ? sequence++ : null,
                uri, ResolveSourceUrl(sourceUrl, uri)));
        }
        foreach (Match match in DashBaseUrl.Matches(text))
        {
            var uri = match.Groups[2].Value.Trim();
            if (uri.Length > 0)
            {
                references.Add(new ManifestReference("BASE", null, uri,
                    ResolveSourceUrl(sourceUrl, uri)));
            }
        }
        return new ParsedManifest(DynamicMpd.IsMatch(text), references);
    }

    private static string RewriteHls(string sourceUrl, string text, string assetBasePath)
    {
        var lines = Lines(text);
        for (var index = 0; index < lines.Length; index++)
        {
            var rewritten = HlsUriAttribute.Replace(lines[index], match =>
            {
                var original = First(match.Groups[2].Value, match.Groups[3].Value);
                return $"URI=\"{AssetUrl(assetBasePath, ResolveSourceUrl(sourceUrl, original))}\"";
            });
            var trimmed = rewritten.Trim();
            if (trimmed.Length > 0 && !trimmed.StartsWith('#'))
            {
                rewritten = AssetUrl(assetBasePath, ResolveSourceUrl(sourceUrl, trimmed));
            }
            lines[index] = rewritten;
        }
        return string.Join('\n', lines);
    }

    private static string RewriteDash(string sourceUrl, string text, string assetBasePath)
    {
        var rewritten = DashUriAttribute.Replace(text, match =>
        {
            var original = First(match.Groups[3].Value, match.Groups[4].Value);
            return $"{match.Groups[1].Value}=\"{XmlEscape(AssetUrl(assetBasePath, ResolveSourceUrl(sourceUrl, original)))}\"";
        });
        return DashBaseUrl.Replace(rewritten, match =>
        {
            var prefix = match.Groups[1].Success ? match.Groups[1].Value : string.Empty;
            var value = XmlEscape(AssetUrl(assetBasePath,
                ResolveSourceUrl(sourceUrl, match.Groups[2].Value.Trim())));
            return $"<BaseURL{prefix}>{value}</BaseURL>";
        });
    }

    private static string AssetUrl(string basePath, string sourceUrl) =>
        basePath + "?url=" + Uri.EscapeDataString(sourceUrl).Replace("%24", "$", StringComparison.OrdinalIgnoreCase);

    private static bool IsProgressivePath(string path) =>
        new[] { ".mp4", ".webm", ".mkv", ".mov", ".m4v", ".mp3", ".m4a", ".ogg", ".opus", ".wav", ".flac" }
            .Any(extension => path.EndsWith(extension, StringComparison.Ordinal));

    private static bool IsSegmentPath(string path) =>
        new[] { ".ts", ".m4s", ".cmfv", ".cmfa", ".aac", ".vtt", ".key" }
            .Any(extension => path.EndsWith(extension, StringComparison.Ordinal));

    private static string[] Lines(string text) => Regex.Split(text, "\\r\\n|\\n|\\r");
    private static string PathOnly(string? sourceUrl) => sourceUrl?.Split('?', 2)[0] ?? string.Empty;
    private static string MediaType(string? contentType) =>
        contentType?.Split(';', 2)[0].Trim().ToLowerInvariant() ?? string.Empty;
    private static string First(string first, string second) => first.Length > 0 ? first : second;
    private static string XmlEscape(string value) => WebUtility.HtmlEncode(value).Replace("'", "&apos;", StringComparison.Ordinal);

    public sealed record ContentRange(long Start, long End, long? Total);
    public sealed record ManifestReference(string RelationType, long? Sequence, string OriginalUri,
        string ResolvedSourceUrl);
    public sealed record ParsedManifest(bool Live, IReadOnlyList<ManifestReference> References);
}
